instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/mongo.go

```diff
@@ -32,25 +33,28 @@ type monitor struct {
 }
 
 func (m *monitor) Started(ctx context.Context, evt *event.CommandStartedEvent) {
-	var spanName string
-
 	hostname, port := peerInfo(evt)
 
-	attrs := []attribute.KeyValue{
-		semconv.DBSystemMongoDB,
-		semconv.DBOperation(evt.CommandName),
-		semconv.DBName(evt.DatabaseName),
-		semconv.NetPeerName(hostname),
-		semconv.NetPeerPort(port),
-		semconv.NetTransportTCP,
-	}
+	attrs := []attribute.KeyValue{semconv.DBSystemMongoDB}
+
+	attrs = appendOpNameAttrs(attrs, evt.CommandName)
+	attrs = appendDBNamespace(attrs, evt.DatabaseName)
+	attrs = appendNetworkPort(attrs, port)
+	attrs = appendNetworkHost(attrs, hostname)
+	attrs = appendNetworkAddress(attrs, net.JoinHostPort(hostname, strconv.Itoa(port)))
+	attrs = appendNetworkTransport(attrs)
```

Comment:
Rather than creating one method for each attribute, how about one method which returns all attributes, and can decide their key name.
Kind of like what the `otelhttp` library does, where we define two structs, one for the old and another for the new version.
https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/net/http/otelhttp/internal/semconv

I would also move the logic into an internal package, for a better separation of concerns.

Reply:
That's an interesting idea. If I understand correctly, you're suggesting we define two separate command monitors specifically for the semantic conventions. Could you explain the benefits of this pattern over the more atomic append* solution proposed in the PR? The goal of the existing solution is to minimize code changes when advancing the library to a desired version, while also avoiding duplication in the monitoring logic.

Reply:
Your solution works if some values are renamed. But if a new value is added, or one is removed then, it becomes trickier to handle since the root method has to know about everything.

With one struct for each semconv version, that struct knows exactly what to do for that version, and that's it. There aren't two versions mixed.
It would also avoid having to define one method for each attribute.

Finally, doing the separation this way also makes it easier to add a third version if we someday need it (I really hope we don't).

Reply:
At last using the same patterns in multiple modules help in maintenance of the repository.

Reply:
@dmathieu Here are my arguments for the factory pattern over struct duplication:

> if a new value is added, or one is removed then, it becomes trickier to handle since the root method has to know about everything.

In this case you would just add `nil` to the unused version, `networkHost` for example:

```go
func appendNetworkHost(attrs []attribute.KeyValue, h string) []attribute.KeyValue {
	return appendAttrs(attrs, semconv1210.NetPeerName, nil, h)
}
```

> It would also avoid having to define one method for each attribute.

This is pretty low complexity, exchanging 1 line of code for 3 to avoid refactoring the library.

> Finally, doing the separation this way also makes it easier to add a third version if we someday need it (I really hope we don't).

By design, there can never be a third version. The clarification of this point can be found here: https://github.com/open-telemetry/semantic-conventions/issues/1502. Additionally, this logic should be rolled back entirely once we release a stable version:

> we only considered using OTEL_SEMCONV_STABILITY_OPT_IN for the initial unstable to stable migration, and not subsequent major version bumps from stable -> (breaking) stable

Reply:
@prestonvasquez, please first create a PR which refactors `otelhttp`. We need to try using similar patterns across the codebase to make it easier to maintain and more readable.

Reply:
@pellared I'm concerned a bug could be introduced by refactoring so much code. I've updated otelmongo to use the otelhttp pattern.